crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -658,70 +643,75 @@ where
         ValidateResult::UnresolvedDependencies(UnresolvedDependencies::Hashes(hashes)) => {
-            // This is the base case where we've been recursing and start seeing
-            // all the same hashes unresolved that we already tried to fetch.
-            // At this point we should just give up on the inline recursing and
-            // let some future background task attempt to fetch these hashes
-            // again. Hopefully by then the hashes are fetchable.
-            // 20 is a completely arbitrary max recursion depth.
-            if recursion_depth > 20 || hashes.iter().all(|hash| fetched_deps.contains(hash)) {
-                Ok(Outcome::AwaitingDeps(hashes))
-            } else {
-                let in_flight = hashes.into_iter().map(|hash| async {
-                    let cascade_workspace = workspace_read.clone();
-                    let cascade = CascadeImpl::from_workspace_and_network(
-                        &cascade_workspace,
-                        network.clone(),
-                    );
-                    cascade
-                        .fetch_record(hash.clone(), NetworkGetOptions::must_get_options())
-                        .await?;
-                    Ok(hash)
-                });
-                let results: Vec<_> = futures::stream::iter(in_flight)
-                    // 10 is completely arbitrary.
-                    .buffered(10)
-                    .collect()
-                    .await;
-                let results: AppValidationResult<Vec<_>> = results.into_iter().collect();
-                for hash in results? {
-                    fetched_deps.insert(hash);
+            tracing::debug!("unresolved dependencies: hashes {hashes:?}");
+            // fetch all missing hashes in the background without awaiting them
+            let cascade_workspace = workspace_read.clone();
+            let cascade =
+                CascadeImpl::from_workspace_and_network(&cascade_workspace, network.clone());
+            // build a collection of futures to fetch the individual missing
+            // hashes
+            let fetches = hashes.clone().into_iter().map(move |hash| {
+                let cascade = cascade.clone();
+                // keep track of which dependencies are being fetched to
+                // prevent multiple fetches of the same hash
+                let fetched_dependencies = fetched_dependencies.clone();
+                async move {
+                    let new_dependency = fetched_dependencies.lock().insert(hash.clone());
+                    // fetch dependency if it is not being fetched yet
+                    if new_dependency {
+                        let result = cascade
+                            .fetch_record(hash.clone(), NetworkGetOptions::must_get_options())
+                            .await;
+                        if let Err(err) = result {
+                            tracing::warn!("error fetching dependent hash {hash:?}: {err}");
```

Comment:
It seems like more should happen here? This hash will just stay in the `fetched_dependencies` set so we won't try to fetch it again and the op that needed it cannot ever pass validation.

Reply:
Either the parent op is rejected here or it needs to be allowed to retry I think. Again, maybe something for the follow up but if it's being left, I think a TODO here might be good

Reply:
Great you pointed that out. I opted for simply removing the hash from the tracked deps collection. That'll lead to another fetch attempt during the subsequent workflow run and is the simplest solution. I see that previously an error led to op validation failure. I think this is an improvement now.